modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj

```diff
@@ -386,10 +387,16 @@
   mbql.u/dispatch-by-clause-name-or-class)
 
 (defmethod compile-filter :between
-  [[_ field min-val max-val]]
-  (compile-filter [:and
-                   [:>= field min-val]
-                   [:<= field max-val]]))
+  [[_ field min-val max-val :as expr]]
+  (if (= :+ (first field))
+    ;; filter with offset
+    (let [[_ field [_ amount unit]] field]
```

Comment:
Probably not a good idea to assume every `[:+ ...]` has a `:field` and an `:interval` as its args at all. Here's a counter example

```clj
[:between [:+ [:field 1 nil] 2] ...]
```

or even

```clj
[:between [:+ [:interval -3 :day] [:field 2 nil]]]
```

Reply:
Probably also a bad idea to assume `min-val` and `max-val` are integers. What if they're timestamps or something like that?

```clj
[:between
 [:+ [:field 1 {:temporal-unit :month}] [:interval -1 :day]]
 [:absolute-datetime #t "2022-06-13"]]

Reply:
Either way this sort of transformation is not specific to MongoDB so it would make sense to do this as part of general query middleware so it applies to all databases. That would make it a lot easier to test as well.

Reply:
I completely agree that this is code is very brittle that's why I'm looking for advice where and how to do it more generally.
I'm a bit confused as to what you suggest. I had the impression that Mongo wasn't supporting an expression as the first argument of `:between`, that's why my tests failed. So I figured adding this minimal support won't make things worse. Are you suggesting we should make the scope larger and implement general support for expressions instead of just fields? And implement this transformation not just for Mongo but other drivers as well? (Postgres and MySQL don't seem to need, for example)).

Reply:
Our MongoDB driver does support expressions as long as the DB is MongoDB 4.0 or above

https://github.com/metabase/metabase/blob/master/modules/drivers/mongo/src/metabase/driver/mongo.clj#L226-L231

Reply:
You are right that the tests are failing because MongoDB isn't supporting expressions as arguments to `:between` (or other filter clauses), which should be allowed; however you don't need to "implement general support for expressions" since we already have that

Reply:
After looking at this a bit more I think you just need to make sure we recursively call the appropriate compilation method on the args to the filters. I think `->rvalue` is what you want here. That's how the SQL QP version, `->honeysql`, does things. You just need to do that here too. e.g. recursively compile `field` (sort of a misnomer since it could be an expression or something else) so we get `{"$add" ...}` instead. Give that a try.

You could alternatively write a whole-query transformation to optimize that temporal arithmetic out entirely but that's more complicated and less general.

Reply:
Also it looks like this bug probably affects other types of filter clauses as well such as `:<` or `:>=`. So you should write tests for those and make sure those get fixed as part of this PR

Reply:
I'm confused. Why doesn't it pass the tests then?

Reply:
Just transforming `field` with `->rvalue` doesn't currently work: it doesn't support translating `:interval`s. Now, I could try adding that, but the Mongo 4.2 docs say

> Adds numbers together or adds numbers and a date. If one of the arguments is a date, [$add](https://www.mongodb.com/docs/v4.2/reference/operator/aggregation/add/#exp._S_add) treats the other arguments as milliseconds to add to the date.

That makes translating `[:interval -3 :year]` on it's own impossible, the number of milliseconds we have to add depends on the date we are adding to. I have no idea at all how to translate that to milliseconds if we are adding to a field.

Reply:
I think using `$dateAdd` is fine. We can throw an Exception if the Database version isn't 5.0+ advising people that we only support it for Mongo 5 or above. It's a tiny edge case we didn't know about until now, so I think that's an acceptable solution. 

You'll just have to figure out how to only run that test for Mongo if we're testing against 5 or above... `metabase.driver/describe-database` is supposed to return `:version` so you can write some sort of logic that only runs the test for Mongo if version is >= 5.0. A little hacky but we don't have a better way of doing that at this point in time and since it's a one-off I think it would be fine as long as you write down why we're doing it.

Reply:
@camsaul At what point should the DB version be checked? Is it available in some context (I think I've seen some kind of store "store") during compilation? Can I delay it until the point when we get to compiling the date arithmetic?

Reply:
Yeah I think doing the check when compiling the arithmetic is fine. `(qp.store/db)` will give you the current Database that we're running the query against so you can pass that in to `driver/details` or whatever you're using to get the version